Mysql build error

## Problem description

The following error occurs building on MacOS:

~~~
ld: warning: dylib (/usr/local/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/lib/python3.10/config-3.10-darwin/libpython3.10.dylib) was built for newer macOS version (13.0) than being linked (10.15.4)
  CXXLD    module/mysql/mysql.la
ld: library not found for -lzstd
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [module/mysql/mysql.la] Error 1
make: *** [install] Error 2
ERROR [build.sh]: unable to make build`
~~~

--- a/configure.ac
+++ b/configure.ac
@@ -519,10 +519,11 @@ AS_IF([test "x$MYSQL_CONFIG" = "xno"],
 AS_IF([test "x$MYSQL" != "xno"],
     [
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS=$(mysql_config --cflags)
+            MYSQL_CPPFLAGS="$(mysql_config --cflags) -DHAVE_MYSQL"
             MYSQL_LDFLAGS=$(mysql_config --libs)
             MYSQL_LIBS=$(mysql_config --libs)
             HAVE_MYSQL=yes
+            GLD_CPPFLAGS="$GLD_CPPFLAGS -DHAVE_MYSQL"
 
             # Define HAVE_MYSQL
             AC_DEFINE([HAVE_MYSQL], [1], [Define when you have MySQL])

--- a/configure.ac
+++ b/configure.ac
@@ -519,9 +519,12 @@ AS_IF([test "x$MYSQL" != "xno"],
         # Check if the mysql_config tool is found
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="`$MYSQL_CONFIG --cflags`"
-            MYSQL_LDFLAGS="`$MYSQL_CONFIG --libs`"
-            MYSQL_LIBS="`$MYSQL_CONFIG --libs`"
+            MYSQL_CPPFLAGS="$MYSQL_CONFIG --cflags"
+            MYSQL_LDFLAGS="$MYSQL_CONFIG --libs"
+            AS_CASE([$host],
+                [*mingw*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"],
+                [*linux*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"],
+                [*darwin*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"])
 
             # Print the result
             echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"

--- a/configure.ac
+++ b/configure.ac
@@ -508,8 +508,13 @@ AS_IF([test -n "$DX_DOXYGEN"],
       [HAVE_DOXYGEN=no])
 AM_CONDITIONAL([HAVE_DOXYGEN], [test ! "x$HAVE_DOXYGEN" = "xno"])
 
-# Check for mysql
-AC_CHECK_LIB([mysqlclient], [mysql_init], [MYSQL="yes"], [MYSQL="no"])
+# Check for mysql_config tool
+AC_PATH_PROG(MYSQL_CONFIG, mysql_config, no)
+
+AS_IF([test "x$MYSQL_CONFIG" = "xno"],
+    [MYSQL="no"
+     AC_MSG_WARN([mysql_config not found - MySQL support will be disabled])],
+    [MYSQL="yes"])
 
 AS_IF([test "x$MYSQL" != "xno"],
     [
@@ -519,12 +524,12 @@ AS_IF([test "x$MYSQL" != "xno"],
         # Check if the mysql_config tool is found
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="$MYSQL_CONFIG --cflags"
-            MYSQL_LDFLAGS="$MYSQL_CONFIG --libs"
+            MYSQL_CPPFLAGS="x$MYSQL_CONFIG --cflags"
+            MYSQL_LDFLAGS="x$MYSQL_CONFIG --libs"
             AS_CASE([$host],
-                [*mingw*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"],
-                [*linux*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"],
-                [*darwin*], [MYSQL_LIBS="$MYSQL_CONFIG --libs"])
+                [*mingw*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"],
+                [*linux*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"],
+                [*darwin*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"])
 
             # Print the result
             echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"

--- a/configure.ac
+++ b/configure.ac
@@ -518,26 +518,13 @@ AS_IF([test "x$MYSQL_CONFIG" = "xno"],
 
 AS_IF([test "x$MYSQL" != "xno"],
     [
-        # Use AC_PATH_TOOL to find the mysql_config tool
-        AC_PATH_TOOL([MYSQL_CONFIG], [mysql_config], [no])
-
-        # Check if the mysql_config tool is found
-        if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="$MYSQL_CONFIG --cflags"
-            MYSQL_LDFLAGS="$MYSQL_CONFIG --libs"
-            MYSQL_LIBS="$MYSQL_CONFIG --libs"
+            MYSQL_CPPFLAGS=$(mysql_config --cflags)
+            MYSQL_LDFLAGS=$(mysql_config --libs)
+            MYSQL_LIBS=$(mysql_config --libs)
 
-            # Print the result
-            echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"
-            echo "MYSQL_LIBS: $MYSQL_LIBS"
-            HAVE_MYSQL=yes
-
-            # Set the paths based on mysql_config output
-            AC_DEFINE([HAVE_MYSQL], [1], [Define if you have MySQL])
-        else
-            AC_MSG_ERROR([mysql_config not found. Please install MySQL development package.])
-        fi
+            # Define HAVE_MYSQL
+            AC_DEFINE([HAVE_MYSQL], [1], [Define when you have MySQL])
     ],
     [HAVE_MYSQL=no])
 

--- a/configure.ac
+++ b/configure.ac
@@ -519,26 +519,9 @@ AS_IF([test "x$MYSQL" != "xno"],
         # Check if the mysql_config tool is found
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="-I`$MYSQL_CONFIG --cflags | sed 's/^-I//; s/ .*//'`"
-            # Use mysql_config to get include and library directories
-            MYSQL_CONFIG_LIBS="$(mysql_config --libs)"
-            MYSQL_LDFLAGS=""
-            MYSQL_LIBS=""
-
-            # Split the input string by spaces
-            IFS=' ' items=($MYSQL_CONFIG_LIBS)
-
-            # Iterate through the items and add to MYSQL_LDFLAGS or MYSQL_LIBS
-            for item in ${items[*]}; do
-            case $item in
-                -L*)
-                MYSQL_LDFLAGS+=" $item"
-                ;;
-                -l*)
-                MYSQL_LIBS+=" $item"
-                ;;
-            esac
-            done
+            MYSQL_CPPFLAGS="$(mysql_config --cflags)"
+            MYSQL_LDFLAGS="$(mysql_config --libs)"
+            MYSQL_LIBS="$(mysql_config --libs)"
 
             # Print the result
             echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"

--- a/configure.ac
+++ b/configure.ac
@@ -519,9 +519,9 @@ AS_IF([test "x$MYSQL" != "xno"],
         # Check if the mysql_config tool is found
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="$(mysql_config --cflags)"
-            MYSQL_LDFLAGS="$(mysql_config --libs)"
-            MYSQL_LIBS="$(mysql_config --libs)"
+            MYSQL_CPPFLAGS="`$MYSQL_CONFIG --cflags`"
+            MYSQL_LDFLAGS="`$MYSQL_CONFIG --libs`"
+            MYSQL_LIBS="`$MYSQL_CONFIG --libs`"
 
             # Print the result
             echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"

--- a/configure.ac
+++ b/configure.ac
@@ -520,11 +520,26 @@ AS_IF([test "x$MYSQL" != "xno"],
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
             MYSQL_CPPFLAGS="-I`$MYSQL_CONFIG --cflags | sed 's/^-I//; s/ .*//'`"
-            MYSQL_LDFLAGS="-L`$MYSQL_CONFIG --libs | sed 's/^-L//; s/ .*//'`"
-            AS_CASE([$host],
-                [*mingw*], [MYSQL_LIBS="-mysql -lzstd -lssl -lcrypto -lresolv -lm"],
-                [*linux*], [MYSQL_LIBS="-lmysqlclient -lzstd -lssl -lcrypto -lresolv -lm"],
-                [*darwin*], [MYSQL_LIBS="-lmysqlclient -lz -lzstd -lssl -lcrypto -lresolv"])
+            # Use mysql_config to get include and library directories
+            MYSQL_CONFIG_LIBS="$(mysql_config --libs)"
+            MYSQL_LDFLAGS=""
+            MYSQL_LIBS=""
+
+            # Split the input string by spaces
+            IFS=' ' read -ra items <<< "$MYSQL_CONFIG_LIBS"
+
+            # Iterate through the items and add to MYSQL_LDFLAGS or MYSQL_LIBS
+            for item in "${items[@]}"; do
+            if [[ $item == -L* ]]; then
+                MYSQL_LDFLAGS+=" $item"
+            elif [[ $item == -l* ]]; then
+                MYSQL_LIBS+=" $item"
+            fi
+            done
+
+            # Print the result
+            echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"
+            echo "MYSQL_LIBS: $MYSQL_LIBS"
             HAVE_MYSQL=yes
 
             # Set the paths based on mysql_config output

--- a/configure.ac
+++ b/configure.ac
@@ -524,7 +524,7 @@ AS_IF([test "x$MYSQL" != "xno"],
             AS_CASE([$host],
                 [*mingw*], [MYSQL_LIBS="-mysql -lzstd -lssl -lcrypto -lresolv -lm"],
                 [*linux*], [MYSQL_LIBS="-lmysqlclient -lzstd -lssl -lcrypto -lresolv -lm"],
-                [*darwin*], [MYSQL_LIBS="-lmysqlclient -lzstd -lssl -lcrypto -lresolv -lm"])
+                [*darwin*], [MYSQL_LIBS="-lmysqlclient -lz -lzstd -lssl -lcrypto -lresolv"])
             HAVE_MYSQL=yes
 
             # Set the paths based on mysql_config output

--- a/configure.ac
+++ b/configure.ac
@@ -522,6 +522,7 @@ AS_IF([test "x$MYSQL" != "xno"],
             MYSQL_CPPFLAGS=$(mysql_config --cflags)
             MYSQL_LDFLAGS=$(mysql_config --libs)
             MYSQL_LIBS=$(mysql_config --libs)
+            HAVE_MYSQL=yes
 
             # Define HAVE_MYSQL
             AC_DEFINE([HAVE_MYSQL], [1], [Define when you have MySQL])

--- a/configure.ac
+++ b/configure.ac
@@ -524,12 +524,9 @@ AS_IF([test "x$MYSQL" != "xno"],
         # Check if the mysql_config tool is found
         if test "x$MYSQL_CONFIG" != "xno"; then
             # Use mysql_config to get include and library directories
-            MYSQL_CPPFLAGS="x$MYSQL_CONFIG --cflags"
-            MYSQL_LDFLAGS="x$MYSQL_CONFIG --libs"
-            AS_CASE([$host],
-                [*mingw*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"],
-                [*linux*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"],
-                [*darwin*], [MYSQL_LIBS="x$MYSQL_CONFIG --libs"])
+            MYSQL_CPPFLAGS="$MYSQL_CONFIG --cflags"
+            MYSQL_LDFLAGS="$MYSQL_CONFIG --libs"
+            MYSQL_LIBS="$MYSQL_CONFIG --libs"
 
             # Print the result
             echo "MYSQL_LDFLAGS: $MYSQL_LDFLAGS"

--- a/configure.ac
+++ b/configure.ac
@@ -526,15 +526,18 @@ AS_IF([test "x$MYSQL" != "xno"],
             MYSQL_LIBS=""
 
             # Split the input string by spaces
-            IFS=' ' read -ra items <<< "$MYSQL_CONFIG_LIBS"
+            IFS=' ' items=($MYSQL_CONFIG_LIBS)
 
             # Iterate through the items and add to MYSQL_LDFLAGS or MYSQL_LIBS
-            for item in "${items[@]}"; do
-            if [[ $item == -L* ]]; then
+            for item in ${items[*]}; do
+            case $item in
+                -L*)
                 MYSQL_LDFLAGS+=" $item"
-            elif [[ $item == -l* ]]; then
+                ;;
+                -l*)
                 MYSQL_LIBS+=" $item"
-            fi
+                ;;
+            esac
             done
 
             # Print the result

--- a/module/mysql/collector.cpp
+++ b/module/mysql/collector.cpp
@@ -155,10 +155,10 @@ int collector::init(OBJECT *parent)
 			size_t n;
 			for ( n=0 ; n<n_aggregates ; n++ )
 			{
-				snprintf(buffer+eos,sizeof(buffer)-eos-1"`%s` double, ",names[n]);
+				snprintf(buffer+eos,sizeof(buffer)-eos-1,"'%s' double, ",names[n]);
 				eos = strlen(buffer);
 			}
-			snprintf(buffer+eos,sizeof(buffer)-eos-1"%s","INDEX i_t (t))");
+			snprintf(buffer+eos,sizeof(buffer)-eos-1,"%s","INDEX i_t (t))");
 			eos = strlen(buffer);
 
 			if ( !db->query(buffer) )
@@ -217,17 +217,17 @@ TIMESTAMP collector::commit(TIMESTAMP t0, TIMESTAMP t1)
 		size_t n;
 		for ( n=0 ; n<n_aggregates ; n++ )
 		{
-			snprintf(buffer+eos,sizeof(buffer)-eos-1",`%s`",names[n]);
+			snprintf(buffer+eos,sizeof(buffer)-eos-1,",`%s`",names[n]);
 			eos = strlen(buffer);
 		}
-		snprintf(buffer+eos,sizeof(buffer)-eos-1") VALUES (from_unixtime(%lli)",db->convert_to_dbtime(gl_globalclock));
+		snprintf(buffer+eos,sizeof(buffer)-eos-1,") VALUES (from_unixtime(%lli)	",db->convert_to_dbtime(gl_globalclock));
 		eos = strlen(buffer);
 		for ( n=0 ; n<n_aggregates ; n++ )
 		{
-			snprintf(buffer+eos,sizeof(buffer)-eos-1",%g",list[n].get_value());
+			snprintf(buffer+eos,sizeof(buffer)-eos-1,",%g",list[n].get_value());
 			eos = strlen(buffer);
 		}
-		snprintf(buffer+eos,sizeof(buffer)-eos-1"%s",")");
+		snprintf(buffer+eos,sizeof(buffer)-eos-1,"%s",")");
 		eos = strlen(buffer);
 
 		if ( !db->query(buffer) )

--- a/module/mysql/database.cpp
+++ b/module/mysql/database.cpp
@@ -107,7 +107,7 @@ int database::init(OBJECT *parent)
 	}
 
 	// set options
-	my_bool report_truncate = false;
+	bool report_truncate = false;
 	if ( mysql_options(mysql_client,MYSQL_REPORT_DATA_TRUNCATION,&report_truncate)!=0 )
 		gl_warning("unable to disable data truncation reporting in mysql");
 
@@ -303,10 +303,10 @@ const char *database::get_last_error(void)
 {
 	return mysql_error(mysql);
 }
-bool database::get_sqlbind(MYSQL_BIND &value, gld_property &target, my_bool *error)
+bool database::get_sqlbind(MYSQL_BIND &value, gld_property &target, bool *error)
 {
 	memset(&value,0,sizeof(value));
-	switch ( target.get_type() ) {
+	switch ( target.get_type() ) {	
 	case PT_double:
 	case PT_random:
 	case PT_loadshape:
@@ -336,6 +336,7 @@ char *database::get_sqldata(char *buffer, size_t size, gld_property &prop, doubl
 				return NULL;
 			return buffer;
 		}
+		break;
 	default:
 		return NULL;
 	}

--- a/module/mysql/recorder.cpp
+++ b/module/mysql/recorder.cpp
@@ -15,12 +15,12 @@ EXPORT_COMMIT(recorder);
 CLASS *recorder::oclass = NULL;
 recorder *recorder::defaults = NULL;
 
-vector<string> split(char* str, const char* delim)
+std::vector<std::string> split(char* str, const char* delim)
 {
     char* saveptr;
     char* token = strtok_r(str,delim,&saveptr);
 
-    vector<string> result;
+    std::vector<std::string> result;
 
     while(token != NULL)
     {
@@ -74,8 +74,8 @@ int recorder::create(void)
 	db = last_database;
 	strcpy(datetime_fieldname,"t");
 	strcpy(recordid_fieldname,"id");
-	property_target = new vector<gld_property>;
-	property_unit = new vector<gld_unit>;
+	property_target = new std::vector<gld_property>;
+	property_unit = new std::vector<gld_unit>;
 	return 1; /* return 1 on success, 0 on failure */
 }
 
@@ -164,13 +164,13 @@ int recorder::init(OBJECT *parent)
 	}
 
 	// connect the target properties
-	vector<string> property_specs = split(get_property(), ", \t;");
+	std::vector<std::string> property_specs = split(get_property(), ", \t;");
 	char property_list[65536]="";
 	for ( size_t n = 0 ; n < property_specs.size() ; n++ )
 	{
 		char buffer[1024];
 		strcpy(buffer,(const char*)property_specs[n].c_str());
-		vector<string> spec = split(buffer,"[]");
+		std::vector<std::string> spec = split(buffer,"[]");
 		if ( spec.size()>0 )
 		{
 			strcpy(buffer,(const char*)spec[0].c_str());
@@ -237,7 +237,7 @@ int recorder::init(OBJECT *parent)
 			exception("cannot find header fields without a parent");
 		char buffer[1024];
 		strcpy(buffer,header_fieldnames);
-		vector<string> header_specs = split(buffer, ",");
+		std::vector<std::string> header_specs = split(buffer, ",");
 		size_t header_pos = 0;
 		header_data[0] = '\0';
 		for ( size_t n = 0 ; n < header_specs.size() ; n++ )

--- a/setup/darwin_20.sh
+++ b/setup/darwin_20.sh
@@ -23,7 +23,6 @@ if ! python$PYTHON_VERSION -m venv -h 1>/dev/null 2>&1 ; then
     INSTALL brew install python$PYTHON_VERSION-venv
     python$PYTHON_VERSION -m venv -h >/dev/null || error "unable to install python$PYTHON_VERSION-venv"
 fi
-iINSTALL python$PYTHON_VERSION -m pip install setuptools
 
 # create python venv for setup if not already done
 if [ ! -x "$PYTHON_EXEC" ] ; then

--- a/setup/darwin_20.sh
+++ b/setup/darwin_20.sh
@@ -58,6 +58,15 @@ clang -v >/dev/null || error "you have not installed clang. Use 'xcode-select --
 # # update library paths
 # INSTALL ldconfig
 
+# install mysql
+if ! mysql_config --libs >/dev/null 2>&1 ; then
+    printf "Installing MySQL... "
+    brew install mysql
+    if ! mysql_config --libs >/dev/null 2>&1 ; then
+        error "Failed to install MySQL with Homebrew."
+    fi
+fi
+
 # install autoconf 2.71 as required
 if [ "$(autoconf --version | head -n 1 | cut -f4 -d' ')" != "2.71" ] ; then
     (cd /tmp ; curl -sL https://ftpmirror.gnu.org/autoconf/autoconf-2.71.tar.gz | tar xz )

--- a/setup/darwin_21.sh
+++ b/setup/darwin_21.sh
@@ -23,7 +23,6 @@ if ! python$PYTHON_VERSION -m venv -h 1>/dev/null 2>&1 ; then
     INSTALL brew install python$PYTHON_VERSION-venv
     python$PYTHON_VERSION -m venv -h >/dev/null || error "unable to install python$PYTHON_VERSION-venv"
 fi
-iINSTALL python$PYTHON_VERSION -m pip install setuptools
 
 # create python venv for setup if not already done
 if [ ! -x "$PYTHON_EXEC" ] ; then

--- a/setup/darwin_21.sh
+++ b/setup/darwin_21.sh
@@ -58,6 +58,15 @@ clang -v >/dev/null || error "you have not installed clang. Use 'xcode-select --
 # # update library paths
 # INSTALL ldconfig
 
+# install mysql
+if ! mysql_config --libs >/dev/null 2>&1 ; then
+    printf "Installing MySQL... "
+    brew install mysql
+    if ! mysql_config --libs >/dev/null 2>&1 ; then
+        error "Failed to install MySQL with Homebrew."
+    fi
+fi
+
 # install autoconf 2.71 as required
 if [ "$(autoconf --version | head -n 1 | cut -f4 -d' ')" != "2.71" ] ; then
     (cd /tmp ; curl -sL https://ftpmirror.gnu.org/autoconf/autoconf-2.71.tar.gz | tar xz )

--- a/setup/darwin_22.sh
+++ b/setup/darwin_22.sh
@@ -23,7 +23,6 @@ if ! python$PYTHON_VERSION -m venv -h 1>/dev/null 2>&1 ; then
     INSTALL brew install python$PYTHON_VERSION-venv
     python$PYTHON_VERSION -m venv -h >/dev/null || error "unable to install python$PYTHON_VERSION-venv"
 fi
-iINSTALL python$PYTHON_VERSION -m pip install setuptools
 
 # create python venv for setup if not already done
 if [ ! -x "$PYTHON_EXEC" ] ; then

--- a/setup/darwin_22.sh
+++ b/setup/darwin_22.sh
@@ -58,6 +58,15 @@ clang -v >/dev/null || error "you have not installed clang. Use 'xcode-select --
 # # update library paths
 # INSTALL ldconfig
 
+# install mysql
+if ! mysql_config --libs >/dev/null 2>&1 ; then
+    printf "Installing MySQL... "
+    brew install mysql
+    if ! mysql_config --libs >/dev/null 2>&1 ; then
+        error "Failed to install MySQL with Homebrew."
+    fi
+fi
+
 # install autoconf 2.71 as required
 if [ "$(autoconf --version | head -n 1 | cut -f4 -d' ')" != "2.71" ] ; then
     (cd /tmp ; curl -sL https://ftpmirror.gnu.org/autoconf/autoconf-2.71.tar.gz | tar xz )

